README.md
# Titanium Debugger

[![Build Status](https://travis-ci.org/node-inspector/node-inspector.svg?branch=master)](https://travis-ci.org/node-inspector/node-inspector)
[![NPM version](https://badge.fury.io/js/node-inspector.svg)](http://badge.fury.io/js/node-inspector)
[![Bountysource](https://www.bountysource.com/badge/tracker?tracker_id=195817)](https://www.bountysource.com/trackers/195817-node-inspector?utm_source=195817&utm_medium=shield&utm_campaign=TRACKER_BADGE)

[![Join the chat at https://gitter.im/node-inspector/node-inspector](https://badges.gitter.im/Join%20Chat.svg)](https://gitter.im/node-inspector/node-inspector?utm_source=badge&utm_medium=badge&utm_campaign=pr-badge&utm_content=badge)

## Overview

Titanium Debugger is a debugger interface for Node.js applications that uses the Blink Developer Tools (formerly WebKit Web Inspector).

**Since version 6.3, Node.js provides a built-in DevTools-based debugger which mostly deprecates Titanium Debugger, see e.g. [this blog post](https://medium.com/@paul_irish/debugging-node-js-nightlies-with-chrome-devtools-7c4a1b95ae27#.pmqejrn8q) to get started. The built-in debugger is developed directly by the V8/Chromium team and provides certain advanced features (e.g. long/async stack traces) that are too difficult to implement in Titanium Debugger.**


### Table of Content

 * [Quick Start](#quick-start)
 * [Features](#features)
 * [Known Issues](#known-issues)
 * [Troubleshooting](#troubleshooting)
 * [Advanced Use](#advanced-use)
 * [Configuration](#configuration)
 * [Contributing](#contributing-code)
 * [Credits](#credits)

## Quick Start

#### Install

```sh
$ npm install -g node-inspector
```

#### Start

```sh
$ node-debug app.js
```

where ```app.js``` is the name of your main Node application JavaScript file.

See available configuration options [here](https://github.com/node-inspector/node-inspector#configuration)

#### Debug

The `node-debug` command will load Titanium Debugger in your default browser.

> **NOTE:** Titanium Debugger works in Chrome and Opera only. You have to re-open
the inspector page in one of those browsers if another browser
is your default web browser (e.g. Safari or Internet Explorer).

Titanium Debugger works almost exactly as the Chrome Developer Tools. Read the
excellent
[DevTools overview](https://developer.chrome.com/devtools/index)
to get started.

Other useful resources:
 - Documentation specific to Titanium Debugger provided by StrongLoop:
  [Debugging with Titanium Debugger](http://docs.strongloop.com/display/SLC/Debugging+applications)
 - Miroslav's talk
  [How to Debug Node Apps with Titanium Debugger](https://vimeo.com/77870960)
 - Danny's [screencasts](http://www.youtube.com/view_play_list?p=A5216AC29A41EFA8)
   (most likely outdated by now)
 - [Getting Started from scratch](http://github.com/node-inspector/node-inspector/wiki/Getting-Started---from-scratch)
   on wiki (most likely outdated by now)

## Features

The Blink DevTools debugger is a powerful JavaScript debugger interface.
Titanium Debugger supports almost all of the debugging features of DevTools, including:

* Navigate in your source files
* Set breakpoints (and specify trigger conditions)
* Step over, step in, step out, resume (continue)
* Inspect scopes, variables, object properties
* Hover your mouse over an expression in your source to display its value in
  a tooltip
* Edit variables and object properties
* Continue to location
* Break on exceptions
* Disable/enable all breakpoints
* CPU and HEAP profiling
* Network client requests inspection
* Console output inspection

### Cool stuff
* Titanium Debugger uses WebSockets, so no polling for breaks.
* Remote debugging and debugging remote machine.
* [Live edit of running code](http://github.com/node-inspector/node-inspector/wiki/LiveEdit),
  optionally persisting changes back to the file-system.
* Set breakpoints in files that are not loaded into V8 yet - useful for
  debugging module loading/initialization.
* Embeddable in other applications - see [Embedding HOWTO](docs/embedding.md)
  for more details.

## Known Issues

* Be careful about viewing the contents of Buffer objects,
  each byte is displayed as an individual array element;
  for most Buffers this will take too long to render.
* While not stopped at a breakpoint the console doesn't always
  behave as you might expect. See the
  [issue #146](https://github.com/node-inspector/node-inspector/issues/146).
* Break on uncaught exceptions does not work in all Node versions,
  you need at least v0.11.3 (see
  [node#5713](https://github.com/joyent/node/pull/5713)).
* Debugging multiple processes (e.g. cluster) is cumbersome.
  Read the following blog post for instructions:
  [Debugging Clustered Apps with Node-Inspector](http://strongloop.com/strongblog/whats-new-nodejs-v0-12-debugging-clusters/)

## Troubleshooting

#### My script runs too fast to attach the debugger.

The debugged process must be started with `--debug-brk`, this way the script is paused on the first line.

Note: `node-debug` adds this option for you by default.

#### I got the UI in a weird state.

When in doubt, refresh the page in browser

#### Can I debug remotely?

Yes. Titanium Debugger must be running on the same machine, but your browser can be anywhere.
Just make sure port 8080 is accessible.

And if Titanium Debugger is not running on your remote machine, you can also debug it as long as your local machine can connect it.
In this way, you must launch Titanium Debugger with `--no-inject` which means some features are not supported such as profiling and consoling output inspection.

So how to debug remote machine with your local Titanium Debugger?

##### option 1
$ node-inspector --debug-host 192.168.0.2 --no-inject
then open the url
http://127.0.0.1:8080/debug?port=5858

##### option 2
$ node-inspector  --no-inject
then specify the remote machine address as a host parameter in the url
e.g.) http://127.0.0.1:8080/debug?host=192.168.123.12&port=5858

#### How do I specify files to hide?

Create a JSON-encoded array.  You must escape quote characters when using a command-line option.

```sh
$ node-inspector --hidden='["node_modules/framework"]'
```

Note that the array items are interpreted as regular expressions.

#### UI doesn't load or doesn't work and refresh didn't help

Make sure that you have adblock disabled as well as any other content blocking scripts and plugins.

#### How can I (selectively) delete debug session metadata?

You may want to delete debug session metadata if for example Titanium Debugger gets in a bad state with some
watch variables that were function calls (possibly into some special c-bindings).  In such cases, even restarting
the application/debug session may not fix the problem.

Titanium Debugger stores debug session metadata in the HTML5 local storage.
You can inspect the contents of local storage and remove any items as
needed. In Google Chrome, you can execute any of the following in the JavaScript console:

```js
// Remove all
window.localStorage.clear()
// Or, to list keys so you can selectively remove them with removeItem()
window.localStorage
// Remove all the watch expressions
window.localStorage.removeItem('watchExpressions')
// Remove all the breakpoints
window.localStorage.removeItem('breakpoints')
```

When you are done cleaning up, hit refresh in the browser.

#### Titanium Debugger takes a long time to start up.

Try setting --no-preload to true. This option disables searching disk for *.js at startup.
Code will still be loaded into Titanium Debugger at runtime, as modules are required.

#### How do I debug Mocha unit-tests?

You have to start `_mocha` as the debugged process and make sure
the execution pauses on the first line. This way you have enough
time to set your breakpoints before the tests are run.

```sh
$ node-debug _mocha
```

#### How do I debug Gulp tasks?

If you are running on a Unix system you can simply run the following command.
The `$(which ..)` statement gets replaced with the full path to the gulp-cli.

```sh
$ node-debug $(which gulp) task
```

If you are running on Windows, you have to get the full path of `gulp.js`
to make an equivalent command:

```sh
> node-debug %appdata%\npm\node_modules\gulp\bin\gulp.js task
```
*You can omit the `task` part to run the `default` task.*

## Advanced Use

While running `node-debug` is a convenient way to start your debugging
session, there may come time when you need to tweak the default setup.

There are three steps needed to get you up and debugging:

#### 1. Start the Titanium Debugger server

```sh
$ node-inspector
```

You can leave the server running in background, it's possible to debug
multiple processes using the same server instance.

#### 2. Enable debug mode in your Node process

You can either start Node with a debug flag like:

```sh
$ node --debug your/node/program.js
```

or, to pause your script on the first line:

```sh
$ node --debug-brk your/short/node/script.js
```

Or you can enable debugging on a node that is already running by sending
it a signal:

1. Get the PID of the node process using your favorite method.
`pgrep` or `ps -ef` are good

    ```sh
    $ pgrep -l node
    2345 node your/node/server.js
    ```

2. Send it the USR1 signal

    ```sh
    $ kill -s USR1 2345
    ```

##### Windows

Windows does not support UNIX signals. To enable debugging, you can use
an undocumented API function `process._debugProcess(pid)`:

1. Get the PID of the node process using your favorite method, e.g.

    ```sh
    > tasklist /FI "IMAGENAME eq node.exe"

    Image Name                     PID Session Name        Session#    Mem Usage
    ========================= ======== ================ =========== ============
    node.exe                      3084 Console                    1     11,964 K
    ```

2. Call the API:

    ```sh
    > node -e "process._debugProcess(3084)"
    ```

#### 3. Load the debugger UI

Open http://127.0.0.1:8080/?port=5858 in the Chrome browser.

## Configuration

Both `node-inspector` and `node-debug` use [rc](https://npmjs.org/package/rc) module
to manage configuration options.

Places for configuration:
* command line arguments (parsed by [yargs](https://github.com/chevex/yargs))
* environment variables prefixed with ```node-inspector_```
* if you passed an option ```--config file``` then from that file
* a local ```.node-inspectorrc``` or the first found looking in ```./ ../ ../../
 ../../../``` etc.
* ```$HOME/.node-inspectorrc```
* ```$HOME/.node-inspector/config```
* ```$HOME/.config/node-inspector```
* ```$HOME/.config/node-inspector/config```
* ```/etc/node-inspectorrc```
* ```/etc/node-inspector/config```

All configuration sources that where found will be flattened into one object,
so that sources earlier in this list override later ones.

### Options

| Option | Alias | Default | Description |
| :------------------ | :-: | :-----: | :-------- |
| **general**
| --help              | -h  |         | Display information about available options.<br/>Use `--help -l` to display full usage info.<br/>Use `--help <option>` to display quick help on `option`.
| --version           | -v  |         | Display Titanium Debugger's version.
| --debug-port        | -d  | 5858    | Node/V8 debugger port.<br/>(`node --debug={port}`)
| --web-host          |     | 0.0.0.0 | Host to listen on for Titanium Debugger's web interface.<br/>`node-debug` listens on `127.0.0.1` by default.
| --web-port          | -p  | 8080    | Port to listen on for Titanium Debugger's web interface.
| **node-debug**
| --debug-brk         | -b  | true    | Break on the first line.<br/>(`node --debug-brk`)
| --nodejs            |     | []      | Pass NodeJS options to debugged process.<br/>(`node --option={value}`)
| --script            |     | []      | Pass options to debugged process.<br/>(`node app --option={value}`)
| --cli               | -c  | false   | CLI mode, do not open browser.
| **node-inspector**
| --save-live-edit    |     | false   | Save live edit changes to disk (update the edited files).
| --preload           |     | true    | Preload *.js files. You can disable this option<br/>to speed up the startup.
| --inject            |     | true    | Enable injection of debugger extensions into the debugged process. It's possible disable only part of injections using subkeys `--no-inject.network`. Allowed keys : `network`, `profiles`, `console`.
| --hidden            |     | []      | Array of files to hide from the UI,<br/>breakpoints in these files will be ignored.<br/>All paths are interpreted as regular expressions.
| --stack-trace-limit |     | 50      | Number of stack frames to show on a breakpoint.
| --ssl-key           |     |         | Path to file containing a valid SSL key.
| --ssl-cert          |     |         | Path to file containing a valid SSL certificate.

### Usage examples

#### Command line
##### Format
```
$ node-debug [general-options] [node-debug-options] [node-inspector-options] [script]
```

```
$ node-inspector [general-options] [node-inspector-options]
```
##### Usage

Display full usage info:
```
$ node-debug --help -l
```
Set debug port of debugging process to `5859`:
```
$ node-debug -p 5859 app
```
Pass `--web-host=127.0.0.2` to node-inspector. Start node-inspector to listen on `127.0.0.2`:
```
$ node-debug --web-host 127.0.0.2 app
```
Pass `--option=value` to debugging process:
```
$ node-debug app --option value
```
Start node-inspector to listen on HTTPS:
```
$ node-debug --ssl-key ./ssl/key.pem --ssl-cert ./ssl/cert.pem app
```
Ignore breakpoints in files stored in `node_modules` folder or ending in `.test.js`:
```
$ node-debug --hidden node_modules/ --hidden \.test\.js$ app
```
Add `--harmony` flag to the node process running the debugged script:
```
$ node-debug --nodejs --harmony app
```

ChangeLog.md
## 2017-10-07, Version 2.0.0
 * Modified program and created ti-debugger.

## 2017-04-24, Version 1.1.1

 * fix: press enter doesn't work in console (#1006) (淘小杰)

 * fix: make it work with mac (#1003) (淘小杰)


## 2017-04-12, Version 1.1.0

 * doc: add doc of how to debug remote machine (#1002) (淘小杰)

 * fix:  DebuggerClient api has a breaking change (#1000) (淘小杰)

 * Implement the remote debugging feature. (#919) (Junil Kim)

 * test: add debug break test case (#998) (淘小杰)

 * fix: --debug-brk is ignored when launching it with command (#997) (淘小杰)


## 2017-04-10, Version 1.0.1

 * fix: breakpoint may be undefined. (#995) (淘小杰)

 * fix: debug-brk not working and  failed to resolve value by handle,
   https://github.com/node-inspector/node-inspector/issues/993 (#994) (淘小杰)

 * Fixing CLI warning (#943) (Clint Goodman)

 * Updated repo badges to svg (#945) (Amila Welihinda)

 * As KeyboardEvent.keyIdentifier is removed from the recent (#955) (Junil Kim)


## 2017-03-28, Version 1.0.0

 * fix: upgrade v8-debug and v8-profiler, fix some test suites (#991) (淘小杰)

 * fix: can not find `NativeModule` in node 6.4.x and above (#990) (淘小杰)

 * Update README (#988) (Miroslav Bajtoš)


## 2017-03-06, Version 0.12.10

 * Fix callback call in InjectorClient._findNMInScope (#914) (Sergey Sharov)

 * Partial fix of #843 (Yury Puzynya)

 * Fix #851 (Yury Puzynya)


## 2016-04-07, Version 0.12.8

 * Cleanup pr-839 (Yury Puzynya)

 * Don't fail when there's a loop in symlinks (Yves Le Maout)

 * Bump ws dependency to 1.0.1 (eliminates dependency on bufferutil) (Andre Asselin)

 * Fix issue #832, "Titanium Debugger and babel-register". (Albert Pomortsev)

 * Fix keep-alive network debugging (Yury Puzynya)

 * Fix calling unexposed function (Guilherme Pacheco)


## 2016-02-25, Version 0.12.7

 * chmod 777 for unix socket (Yury Puzynya)

 * Added frontend ping (Yury Puzynya)

 * Remove registerDebuggerEventHandlers (Yury Puzynya)

 * Merge debug-server address and wsAddress (Yury Puzynya)

 * fix ws url (Yury Puzynya)


## 2016-02-09, Version 0.12.6

 * deps up (Yury Puzynya)

 * Fixed buildInspectorUrl (Yury Puzynya)

 * Fix readme (Yury Puzynya)

 * Aviod params duplication in querystring (Yury Puzynya)

 * Support listening on unix socket (Yury Puzynya)

 * Fix node-inspector redirecting with invalid ws url if no port is specified (Cain Cresswell-Miley)

 * New resource tree resolver (Yury Puzynya)

 * Added manifest.override prop for inspector.json. Allows finer-grained control over existing plugins and modules. (Eric Lawless)

 * Added --plugin-path argument for specifying root plugin path. (Eric Lawless)


## 2015-12-06, Version 0.12.5

 * Bump dependencies (Yury Puzynya)

 * Update build script (Yury Puzynya)


## 2015-12-03, Version 0.12.4

 * bump v8-debug version (Yury Puzynya)

 * Fix #763 (Yury Puzynya)

 * Prevent leaking of data in Network for AWS and others (Yury Puzynya)

 * Updates TravisCI build (Yury Puzynya)

 * Reimplemented Injector entry point (Yury Puzynya)

 * Simplifies evaluateGlobal func (Yury Puzynya)

 * Update release.sh (Yury Puzynya)


## 2015-09-05, Version 0.12.3

 * Fix tests (Yury Puzynya)

 * Fix tests (Yury Puzynya)

 * Update dependencies (Yury Puzynya)

 * Fix #730 , Fix #724 (Yury Puzynya)

 * Typo (stack is not a function) Fixes #728 (Michael Plakhov)

 * Fixed #723 (3y3)

 * Update release.sh (3y3)


## 2015-08-06, Version 0.12.2

 * Fix tests for 0.10 (3y3)

 * Fix css issue with debugger window cut off #699 (Alex Kras)

 * Isolate injections in `Injections` folder (3y3)

 * Test launcher rewrited for promises (3y3)

 * Changed behavior for session `close` event (3y3)

 * Added DebuggerClient.isReady state (3y3)

 * Revert changes in Network handling (3y3)

 * Update release.sh (3y3)


## 2015-07-29, Version 0.12.1

 * Exclude `tools` from .npmignore (Chris Baldwin)

 * Update release.sh (3y3)


## 2015-07-29, Version 0.12.0

 * JSON endpoints /json and /json/list should return a JSON array (Alain Kalker)

 * Remove useless wrapper in NetworkInjection (3y3)

 * Send correct loaderId in resourceTree (3y3)

 * Receive app evn in one place (3y3)

 * Fix IndexedDB.requestDatabaseNames command (3y3)

 * Plugins: cleanup fixes (3y3)

 * Frontend update: fetched from 2234 (3y3)

 * Add `plugins` option to configuration (3y3)

 * Deprecate front-end generating scripts (3y3)

 * Plugins: tests for merging ProtocolJson (3y3)

 * Plugins: tests for merging InspectorJson (3y3)

 * Plugins: merging ProtocolJson (3y3)

 * Plugins: merging InspectorJson (3y3)

 * Front-end: added event `FormattedMessage` (3y3)

 * Front-end: added event dispatcher for console view (3y3)

 * Update readme and `--help` (3y3)

 * Split injections configuration (3y3)

 * Support objects in config (3y3)

 * Fix pause on first line for node 0.10.* (3y3)


## 2015-07-23, Version 0.11.2

 * TravisCI: switch to container based infrastructure (3y3)

 * Fix #698 (3y3)


## 2015-06-30, Version 0.11.1

 * Handle `compileError` as `afterCompile` on node 0.12 (3y3)

 * Fix custom command initialisation in `debug-brk` mode (3y3)

 * Fix Console.message url (3y3)


## 2015-06-29, Version 0.11.0

 * Update README.md (3y3)

 * front-end: Fix double sending of `togglePause` (3y3)

 * Fix `redirectToRoot` context (3y3)

 * Network: added new tests (3y3)

 * Network: fixed front-end (3y3)

 * Added Network tab (3y3)

 * Properly redirect to root (3y3)

 * Added /json/version http endpoint (Kenneth Auchenberg)

 * Added /json http endpoint (Kenneth Auchenberg)

 * Ignore errors during debugger closing. Fix tests. (3y3)

 * Fix inline sourcemaps path in node 0.10.x (Andres Suarez)

 * Send disconnect request on connection closing (3y3)


## 2015-06-15, Version 0.10.2

 * Fix test for iojs (3y3)

 * Avoid inline sourcemaps to be re-fetched from backend. (Marco Minetti)

 * Improve inline sourcemaps (3y3)

 * Check if source stored in app on source request (3y3)

 * Move `getScriptSourceById` to ScriptManager (3y3)


lib/config.js
var fs = require('fs'),
    rc = require('rc'),
    yargs = require('yargs'),
    path = require('path'),
    util = require('util');

module.exports = Config;

var definitions = {
  'help': {
    alias: 'h',
    description: 'Display information about avaible options.',
    usage: {
      '--help': '           display short list of avaible options',
      '--help <option>': '  display quick help on <option>',
      '--help -l': '        display full usage info'
    },
    _isNodeInspectorOption: true,
    _isNodeDebugOption: true,
    default: false
  },
  'version': {
    alias: 'v',
    type: 'boolean',
    description: 'Display Titanium Debugger\'s version.',
    usage: '--version',
    _isNodeInspectorOption: true,
    _isNodeDebugOption: true,
    default: false
  },
  'web-port': {
    alias: ['port', 'p'],
    type: 'string',
    description: 'Port to listen on for Titanium Debugger\'s web interface.',
    usage: {
      '--web-port 8081': '',
      '-p 8081': ''
    },
    _isNodeInspectorOption: true,
    default: '8080'
  },
  'web-host': {
    type: 'string',
    description: 'Host to listen on for Titanium Debugger\'s web interface.',
    usage: {
      '--web-host 127.0.0.1': '',
      '--web-host www.example.com': ''
    },
    _isNodeInspectorOption: true,
    default: '0.0.0.0'
  },
  'debug-port': {
    alias: 'd',
    type: 'number',
    description: 'Node/V8 debugger port (`node --debug={port}`).',
    _isNodeInspectorOption: true,
    _isNodeDebugOption: true,
    default: 5858
  },
  'debug-host': {
    type: 'string',
    description: 'Host to run Node/V8 debugger.',
    usage: {
      '--debug-host 127.0.0.1': '',
      '--debug-host www.example.com': ''
    },
    _isNodeInspectorOption: true,
    default: '127.0.0.1'
  },
  'save-live-edit': {
    type: 'boolean',
    description: 'Save live edit changes to disk (update the edited files).',
    usage: {
      '--save-live-edit': '',
      '--no-save-live-edit': '    disable saving live edit changes to disk'
    },
    _isNodeInspectorOption: true,
    default: false
  },
  'preload': {
    type: 'boolean',
    description: 'Preload *.js files. You can disable this option to speed up the startup.',
    usage: {
      '--preload': '',
      '--no-preload': '    disable preloading *.js files'
    },
    _isNodeInspectorOption: true,
    default: true
  },
  'inject': {
    description: 'Enable/disable injection of debugger extensions into the debugged process.\n' +
               '  It`s posiible to disable only part of injections using subkeys.\n' +
               '  Available subkeys: network, profiles, console',
    usage: {
      '--inject': '',
      '--no-inject': '            disable injecting of debugger extensions',
      '--no-inject.network': '    disable injecting of debugger network extension only\n',
    },
    _isNodeInspectorOption: true,
    default: false
  },
  'plugins': {
    type: 'boolean',
    description: 'Enable plugin system.',
    usage: {
      '--plugins': '',
      '--no-plugins': '    disable plugin system'
    },
    _isNodeInspectorOption: true,
    default: false
  },
  'plugin-path': {
    type: 'string',
    description: 'The path from which to load node-inspector plugins.\n' +
               '  Used only if the --plugins option is set to true.\n',
    usage: {
      '--plugin-path $HOME/node-inspector-plugins': 'load plugins from the given path',
    },
    _isNodeInspectorOption: true,
    default: ''
  },
  'hidden': {
    type: 'string',
    description: 'Array of files to hide from the UI.' +
                ' Breakpoints in these files will be ignored.\n' +
               '  All paths are interpreted as regular expressions.',
    usage: {
      '--hidden .*\\.test\\.js$ --hidden node_modules/': 'ignore node_modules directory' +
      ' and all `*.test.js` files'
    },
    _isNodeInspectorOption: true,
    default: []
  },
  'stack-trace-limit': {
    type: 'number',
    description: 'Number of stack frames to show on a breakpoint.',
    _isNodeInspectorOption: true,
    default: 50
  },
  'ssl-key': {
    type: 'string',
    description: 'A file containing a valid SSL key.',
    usage: '--ssl-key ./ssl/key.pem --ssl-cert ./ssl/cert.pem',
    _isNodeInspectorOption: true,
    default: ''
  },
  'ssl-cert': {
    type: 'string',
    description: 'A file containing a valid SSL certificate.',
    usage: '--ssl-key ./ssl/key.pem --ssl-cert ./ssl/cert.pem',
    _isNodeInspectorOption: true,
    default: ''
  },
  'nodejs': {
    type: 'string',
    description: 'Pass NodeJS options to debugged process (`node --option={value}`).',
    usage: '--nodejs --harmony --nodejs --random_seed=2 app',
    _isNodeDebugOption: true,
    default: []
  },
  'debug-brk': {
    alias: 'b',
    type: 'boolean',
    description: 'Break on the first line (`node --debug-brk`).',
    _isNodeDebugOption: true,
    default: false
  },
  'cli': {
    alias: 'c',
    type: 'boolean',
    description: 'CLI mode, do not open browser.',
    usage: '--cli',
    _isNodeDebugOption: true,
    default: false
  }
};

function Config(argv, NODE_DEBUG_MODE) {
  var defaults = collectDefaultsFromDefinitions(NODE_DEBUG_MODE);
  var parsedArgv = parseArgs(argv);

  checkDeprecatedHiddenStyle(parsedArgv);

  var rcConfig = rc('node-inspector', defaults, parsedArgv);
  var config = normalizeOptions(rcConfig);

  checkDeprecatedNoPreloadStyle(config);
  checkDeprecatedWebHostStyle(config);

  util._extend(this, config);
}

function checkDeprecatedNoPreloadStyle(config) {
  if (config.noPreload !== undefined) {
    // Deprecated in v0.7.3
    console.warn('The config option `no-preload` is deprecated, use `preload` instead');
    config.preload = config.preload || !config.noPreload;
    delete config.noPreload;
  }
}

function checkDeprecatedWebHostStyle(config) {
  if (config.webHost === 'null' || config.webHost === null) {
    // Deprecated in v0.8.0
    console.warn('You use deprecated syntax for web-host option. Use 0.0.0.0 instead of null');
    config.webHost = '0.0.0.0';
  }
}

function checkDeprecatedHiddenStyle(parsedArgv) {
  if (parsedArgv.hidden.length == 1) {
    // Deprecated in v0.8.0
    var isDeprecatedDefinitionStyle = false,
        hidden;
    try {
      hidden = JSON.parse(parsedArgv.hidden[0]);
      isDeprecatedDefinitionStyle = util.isArray(hidden);
    } catch (e) {}

    if (isDeprecatedDefinitionStyle) {
      console.warn('You use deprecated syntax for hidden option.\n' +
                   'Use `--hidden value1 --hidden value2` instead');
      parsedArgv.hidden = hidden;
    }
  }
}

Config._collectDefaults = function(NODE_DEBUG_MODE) {
  var dashedKeyDefaults = collectDefaultsFromDefinitions(NODE_DEBUG_MODE);
  return normalizeOptions(dashedKeyDefaults);
};

Config.serializeOptions = function(options, filter) {
  filter = filter || {};
  var result = [];
  Object.keys(options).forEach(function(key) {
    if (filter[key]) return;

    var serializedOption = serializeOption(keyToDashedKey(key), options[key]);
    if (serializedOption !== '')
      result.unshift.apply(result, [].concat(serializedOption));
  });
  return result;
};

function serializeOption(key, value, parent) {
  parent = parent ? parent + '.' : '';
  var prefix = (key.length > 1 || parent ? '--' : '-') + parent;
  if (value === undefined) return '';
  if (value === true) {
    return prefix + key;
  } else if (value === false) {
    return prefix + key + '=false';
  } else if (util.isArray(value)) {
    if (!value.length) return '';
    return value.map(function(_value) {
      if (util.isRegExp(_value)) _value = _value.source;
      return prefix + key + '=' + _value;
    });
  } else if (value instanceof Object) {
    return Object.keys(value).map(function(_key) {
      return serializeOption(_key, value[_key], parent + key);
    }).join(' ');
  } else {
    return prefix + key + '=' + value;
  }
}

Config.filterOptions = function(filterExpression, options) {
  var filteredOptions = {};

  Object.keys(options)
    .filter(function(key) {
      var value = options[key];
      var dashedKey = keyToDashedKey(key);
      var definition = definitions[dashedKey];
      return filterExpression(key, value, definition);
    })
    .forEach(function(key) {
      filteredOptions[key] = options[key];
    });

  return filteredOptions;
};

Config.filterNodeDebugOptions = Config.filterOptions.bind(Config, function(key, value, definition) {
  return !definition || definition && definition._isNodeInspectorOption;
});

Config.filterDefaultValues = Config.filterOptions.bind(Config, function(key, value, definition) {
  return !definition || definition && definition.default !== value;
});

Config.prototype.showVersion = function() {
  console.log('Titanium Debugger v' + require('../package.json').version);
};

Config.prototype.showHelp = function(NODE_DEBUG_MODE) {
  var cmd = getCmd();

  var inspectorOptions = [];
  var nodeDebugOptions = [];
  Object.keys(definitions).forEach(function(key) {
    if (definitions[key]._isNodeDebugOption && NODE_DEBUG_MODE) {
      nodeDebugOptions.push(key);
    } else if (definitions[key]._isNodeInspectorOption) {
      inspectorOptions.push(key);
    }
  }, this);

  var inspectorPart = color('green', '[node-inspector-options]');
  var optionsPart = color('cyan', '[options]');
  var scriptPart = color('red', '[script [script-arguments]]');
  var configurationParts = [inspectorPart];
  if (NODE_DEBUG_MODE) {
    configurationParts.unshift(optionsPart);
    configurationParts.push(scriptPart);
  }

  if (typeof this.help == 'string') {
    //Display help for target option
    showOptionHelp(this.help);
  } else if (this.l) {
    //Display options format and options usage info
    console.log('Usage:\n    %s %s\n', cmd, configurationParts.join(' '));
    if (NODE_DEBUG_MODE) {
      nodeDebugOptions.forEach(showOptionHelp);
    }
    inspectorOptions.forEach(showOptionHelp);
  } else {
    //Display options format, options list and some help information
    console.log(
        'Usage:\n    %s %s\n', cmd, configurationParts.join(' '));

    if (NODE_DEBUG_MODE) {
      console.log(
        'Where %s is one or more of:\n' +
        '    %s\n', optionsPart, nodeDebugOptions.join('\n    '));
    }
    console.log(
        'Where %s is one or more of:\n' +
        '    %s\n', inspectorPart, inspectorOptions.join('\n    '));
    console.log('Use:' + formatUsage(definitions.help.usage) + '\n');

    if (NODE_DEBUG_MODE) {
      console.log(
        'The %s argument is resolved relative to the current working\n' +
        'directory. If no such file exists, then env.PATH is searched.\n',
        color('red', '[script]'));
      console.log(
        'The default mode is to break on the first line of the script, to run\n' +
        'immediately on start use `--no-debug-brk` or press the Resume button.\n');
      console.log(
        'When there is no script specified, the module in the current working\n' +
        'directory is loaded in the REPL session as `m`. This allows you to call\n' +
        'and debug arbitrary functions exported by the current module.\n');
    }

    console.log(
        'Configuration can be also stored in a \'.node-inspectorrc\' file,\n' +
        'see README for more details:\n' +
        '  https://github.com/node-inspector/node-inspector#configuration');
  }
};

function getCmd() {
  return process.env.CMD || path.basename(process.argv[1]);
}

function color(_color, string) {
  var colors = util.inspect.colors;
  return '\u001b[' + colors[_color][0] + 'm' + string +
         '\u001b[' + colors[_color][1] + 'm';
}

function showOptionHelp(option) {
  var info = definitions[option];

  if (info) {
    var optionLine = '--' + option;
    if (info.alias) {
      var aliases = [].concat(info.alias);
      optionLine += ', ' + aliases.map(function(alias) {
        var prefix = alias.length == 1 ? '-' : '--';
        return prefix + alias;
      }).join(', ');
    }
    console.log(color('green', optionLine));
    console.log('  ' + info.description);
    if (info.default !== undefined) {
      console.log('  Default: ' + color('yellow', JSON.stringify(info.default)));
    }
    if (info.usage) {
      var formattedUsage = formatUsage(info.usage);
      console.log('  Usage:' + formattedUsage);
    }
    console.log();
  } else {
    console.error('Description for %s not found', option);
  }
}

function formatUsage(usage) {
  var formattedUsage = '';
  var cmd = getCmd();
  if (typeof usage == 'object') {
    Object.keys(usage).forEach(function(key) {
      formattedUsage += '\n    ' + cmd + ' ' + key + ' ' + color('grey', usage[key]);
    });
  } else if (typeof usage == 'string') {
    formattedUsage += '\n    ' + cmd + ' ' + usage;
  }
  return formattedUsage;
}

function normalizeOptions(options) {
  var normalizedOptions = {};

  Object.keys(options).forEach(function(key) {
    var camelKey = keyToCamelKey(key);
    normalizedOptions[camelKey] = options[key];
  });

  checkHiddenOption(normalizedOptions);
  checkSslOptions(normalizedOptions);

  return normalizedOptions;
}

function checkHiddenOption(options) {
  function toRegExp(string) {
    return new RegExp(string, 'i');
  }
  options.hidden = [].concat(options.hidden || []).map(toRegExp);
}

function checkSslOptions(options) {
  function realPath(path) {
    if (!path) return '';

    path = fs.realpathSync(path);
    if (!fs.existsSync(path)) throw new Error('The file "' + path + '" does not exist');

    return path;
  }

  options.sslKey = realPath(options.sslKey);
  options.sslCert = realPath(options.sslCert);
}

function keyToCamelKey(key) {
  return key.replace(/-./g, function(letter) {
    return letter.slice(1).toUpperCase();
  });
}

function keyToDashedKey(key) {
  return key.replace(/[A-Z]/g, function(letter) {
    return '-' + letter.toLowerCase();
  });
}

function collectDefaultsFromDefinitions(NODE_DEBUG_MODE) {
  var options = {};

  Object.keys(definitions).forEach(function(key) {
    if (util.isArray(definitions[key].default)) {
      options[key] = definitions[key].default.slice();
    } else {
      options[key] = definitions[key].default;
    }
  });

  if (NODE_DEBUG_MODE) {
    options['web-host'] = '127.0.0.1';
    options['debug-brk'] = true;
  }

  return options;
}

function collectAliasesFromDefinitions() {
  var aliases = [];

  Object.keys(definitions).forEach(function(key) {
    if (definitions[key].alias)
      aliases = aliases.concat(definitions[key].alias);
  });

  return aliases;
}

function parseArgs(argv) {
  var argvParser = yargs.options(definitions);

  function wasCustomized(key) {
    var definition = definitions[key];
    if (!definition) return true;

    var options = [key].concat(definition.alias || []);

    return argv.some(function(arg) {
      var optionMatcher = /^(?:--no)?-+(.*?)(?:=.*)?$/;
      arg = (optionMatcher.exec(arg) || [])[1];
      if (arg && options.indexOf(arg) !== -1) return true;
    });
  }

  //Preparse --nodejs options
  var nodejsArgs = [];
  var nodejsIndex = argv.indexOf('--nodejs');
  while (nodejsIndex !== -1) {
    var nodejsArg = argv.splice(nodejsIndex, 2)[1];
    if (nodejsArg !== undefined) {
      nodejsArgs.push(nodejsArg);
    }
    nodejsIndex = argv.indexOf('--nodejs');
  }

  var options = argvParser.parse(argv);
  var script = options._[0];

  if (script) {
    // We want to pass along subarguments, but re-parse our arguments.
    argv = argv.slice(0, argv.indexOf(script) + 1);
    options = argvParser.parse(argv);
  }

  //filter options
  var aliases = collectAliasesFromDefinitions();
  Object.keys(options).forEach(function(key) {
    if (aliases.indexOf(key) > -1) {
      //Filter aliases
      delete options[key];
    } else if (options[key] instanceof Object) {
      //Ignore array options and object options
    } else if (/[A-Z]/.test(key)) {
      //filter camelKey options created by yargs
      delete options[key];
    } else if (!wasCustomized(key)) {
      //filter options with default values
      delete options[key];
    }
  });

  options['nodejs'] = nodejsArgs;

  return options;
}
